### ISSUE TITLE
Implement function to align all images of a stack

### ISSUE THREAD

**Issue by opatiny** [image-js/image-js-typescript#417]
- we should probably implement the `smartAlign` algorithm that was developed in [profid-padif-rework](https://github.com/zakodium/profid-padif-rework) in `image-js-ts`
- align all images
- crop all images in order to have the same dimension of all the images once aligned

**Comment by opatiny** [comment]
This would be the equivalent of the [`matchAndCrop`](https://github.com/image-js/image-js/blob/8bfbcc524daafc35c35df62f5d18e7679deae7ee/src/stack/transform/matchAndCrop.js) function currently in `image-js`.

**Comment by opatiny** [comment]
We have to add a new property to stacks called

```ts
private readonly translations: Point[];
```

We'll divide the alignment into two different methods
1) `align`: modifies the translations property of the stack to have all the translations that have to be applied to the images to align them
2) `cropCommonArea`: creates a new stack with the cropped images (all the same size). The translations of the new stack are all (0,0).

**Comment by opatiny** [comment]
- The align method should work on images with different sizes.
- The user should specify a maximal x and y displacement.
- The pixels that don't overlap are ignored
- For each translation, the similarity is computed and the position with the minimal similarity is written in the `translations` property.
- All images are aligned on the first image of the stack? -> to be discussed

**Comment by targos** [comment]
Related: if we imagine the stack in 3D, is the first image (`images[0]`) at the bottom or at the top?
In other words, if we look at the stack from above, which image do we see?

**Comment by opatiny** [comment]
> Related: if we imagine the stack in 3D, is the first image (`images[0]`) at the bottom or at the top? In other words, if we look at the stack from above, which image do we see?

I feel like the order of the images should be the order in which you would show them if they were part of a movie. In other words, if you loop on the images, it should be the order in which they appear in the "animation".

**Comment by targos** [comment]
If the mental model of it is a movie, "Stack" seems like the wrong name 😕

**Comment by opatiny** [comment]
Indeed, that is how I considered them, but maybe it's not the best. How would you imagine the stacks?

**Comment by opatiny** [comment]
Wdyt @lpatiny ?
